### PR TITLE
[MTLS] Decode utf-8 url encoded certs when passing from header

### PR DESCRIPTION
**Pull request by anjuchamantha**

### CODE REVIEW

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java [hunk 1]
```diff
@@ -286,6 +292,17 @@ private X509Certificate parseCertificate(String content) throws CertificateExcep
                 .generateCertificate(new ByteArrayInputStream(decoded));
     }
 
+    /**
+     * Check if the string contains percent-encoded sequences.
+     *
+     * @param str input string value.
+     * @return returns true if string is url encoded.
+     */
+    private boolean isUrlEncoded(String str) {
```
Comment (DMHP) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:301]: This is not a URL encode, but a certificate encode.

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java [hunk 2]
```diff
@@ -286,6 +292,17 @@ private X509Certificate parseCertificate(String content) throws CertificateExcep
...
+
+        return str.matches(".*(%[0-9a-fA-F]{2}).*");
```
Comment (DMHP) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:303]: Is it reliable to determine whether a certificate is UTF-8 encoded or not using a regex ? Don't we have any libraries to check this?

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java (hunk 1 above)
Comment (anjuchamantha) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:301]: Certificate is url encoded and we use `URLDecoder.decode` to decode

Comment (RivinduM) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:301]: From the gateway when appending the cert as a header, it does a url encoding. That needs to be decoded prior to the certificate decoding. That's what Anju is trying to do here.

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java (hunk 2 above)
Comment (anjuchamantha) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:303]: Couldn't find any

Comment (DMHP) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:303]: Anyway relying on the regex is not reliable and we can not guarantee on the functionality. We may need to look for an alternative.

Comment (anjuchamantha) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:303]: Will try to base64 decode the cert first (as in the earlier implementation) and if it throws an exception will urlDecode before base64 decoding

Comment (anjuchamantha) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java]: Changed with https://github.com/wso2-extensions/identity-oauth-addons/pull/132/commits/88128bd2abe38aa31609ed5f3eb535b774b50a06

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java [hunk 3]
```diff
@@ -271,21 +273,38 @@ private Optional<X509Certificate> getCertificateFromHeader(HttpServletRequest re
...
-        // Removing all whitespaces and new lines.
-        certBody = certBody.replaceAll("\\s", StringUtils.EMPTY).replace("\\n", StringUtils.EMPTY);
-        byte[] decoded = Base64.getDecoder().decode(certBody);
+        byte[] decoded;
+        String sanitizedCertificate = sanitizeCertificate(content);
+        // First we try to Base64 decode, if it is not decodable, we try to url decode first and then Base64 decode.
+        try {
+            decoded = Base64.getDecoder().decode(sanitizedCertificate);
+        } catch (IllegalArgumentException e) {
+            String urlDecodedContent = URLDecoder.decode(content, StandardCharsets.UTF_8.name());
```
Comment (RivinduM) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:284]: Let's add a debug log and log the content. We'll be able to use it for troubleshooting if the fix doesn't work

Comment (anjuchamantha) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:285]: Added with https://github.com/wso2-extensions/identity-oauth-addons/pull/132/commits/c88da558a28ea1a806741447b328214111e6dde2

Comment (anjuchamantha) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java:285]: Is it safe to debug log the certificate here?

**Review by RivinduM — APPROVED** [review]

**Review by DMHP — APPROVED** [review]

**Pull request merged**